## Numbers

Generate a dictionary where the keys are the numbers between 1 and 20 and the values are the square of the keys.

In [8]:
result = {}

for i in range(1, 21):
    result[i] = i**2

print(result)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


In [9]:
result = {i:i**2 for i in range(1, 21)}

print(result)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


The result is

    {1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


Update the dictionary from above with the given dictionary *additional_dict*

In [10]:
additional_dict = {40: 1600, 41: 1681, 42: 1764}

result.update(additional_dict)

print(result)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400, 40: 1600, 41: 1681, 42: 1764}


The result is

    {1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400, 40: 1600, 41: 1681, 42: 1764}


## NYC marathon winners

Here's a (nested) list of the female NYC marathon winners from wikipedia. https://en.wikipedia.org/wiki/List_of_winners_of_the_New_York_City_Marathon in the format

            [['Year 1', 'Winner 1, 'Country 1', 'Time 1', 'Notes 1'],
             ['Year 2', 'Winner 2, 'Country 2', 'Time 2', 'Notes 2'], 
             ...]]
The follwing cell reads in the list from the file *female_winners.csv* and modifies the list to match the format above. Reading in the file will be covered later.

In [3]:
with open("female_winners.csv") as f:
    winners = f.readlines()

winners = [entry.replace("\n","").split(",") for entry in winners]
winners = winners[1:]
winners[:3]

[['1971', 'Beth Bonner', 'United States', '2:55:22', 'World record'],
 ['1972', 'Nina Kuscsik', 'United States', '3:08:41', ''],
 ['1973', 'Nina Kuscsik', 'United States', '2:57:07', 'Second victory']]

### Winner count
Use a dictionary to count how often each winner won the marathon.

*Hint: Loop through the list. Start with an empty dictionary and add winners as a key-value pair as they appear, so you'll end up with a dictionary like {"winner1":count1, "winner2":count2, ...}.*

In [4]:
winner_count = {}

for w in winners:
    name = w[1]
    if name not in winner_count:
        winner_count[name] = 0
    winner_count[name] += 1

print(winner_count)

{'Beth Bonner': 1, 'Nina Kuscsik': 2, 'Kathrine Switzer': 1, 'Kim Merritt': 1, 'Miki Gorman': 2, 'Grete Waitz': 9, 'Allison Roe': 1, 'Priscilla Welch': 1, 'Ingrid Kristiansen': 1, 'Wanda Panfil': 1, 'Liz McColgan': 1, 'Lisa Ondieki': 1, 'Uta Pippig': 1, 'Tegla Loroupe': 2, 'Anuța Cătună': 1, 'Franziska Rochat-Moser': 1, 'Franca Fiacconi': 1, 'Adriana Fernández': 1, 'Lyudmila Petrova': 1, 'Margaret Okayo': 2, 'Joyce Chepchumba': 1, 'Paula Radcliffe': 3, 'Jeļena Prokopčuka': 2, 'Derartu Tulu': 1, 'Edna Kiplagat': 1, 'Firehiwot Dado': 1, 'Priscah Jeptoo': 1, 'Mary Keitany': 4, 'Shalane Flanagan': 1, 'Joyciline Jepkosgei': 1}


The result is

    {'Beth Bonner': 1,
     'Nina Kuscsik': 2,
     'Kathrine Switzer': 1,
     'Kim Merritt': 1,
     'Miki Gorman': 2,
     'Grete Waitz': 9,
     'Allison Roe': 1,
     'Priscilla Welch': 1,
     'Ingrid Kristiansen': 1,
     'Wanda Panfil': 1,
     'Liz McColgan': 1,
     'Lisa Ondieki': 1,
     'Uta Pippig': 1,
     'Tegla Loroupe': 2,
     'Anuța Cătună': 1,
     'Franziska Rochat-Moser': 1,
     'Franca Fiacconi': 1,
     'Adriana Fernández': 1,
     'Lyudmila Petrova': 1,
     'Margaret Okayo': 2,
     'Joyce Chepchumba': 1,
     'Paula Radcliffe': 3,
     'Jeļena Prokopčuka': 2,
     'Derartu Tulu': 1,
     'Edna Kiplagat': 1,
     'Firehiwot Dado': 1,
     'Priscah Jeptoo': 1,
     'Mary Keitany': 4,
     'Shalane Flanagan': 1,
     'Joyciline Jepkosgei': 1}

### Structure one sublist

Create a dictionary like

    {"winner":value1, "country":value2, "time":value3, "note":value4}

with for the 0th item of the winner list.

*Hint: Start with an empty dictionary. Use the zip() function and the list ["winner","country","time","note"]*

In [6]:
winner_dict_0 = {}

keys = ["winner", "country", "time", "note"]
values = winners[0][1:]

result = {k:v  for k,v in zip(keys,values) }

print(result)

{'winner': 'Beth Bonner', 'country': 'United States', 'time': '2:55:22', 'note': 'World record'}


The result is 

    {'winner': 'Beth Bonner',
     'country': 'United States',
     'time': '2:55:22',
     'note': 'World record'}

### Structure the whole list, nested dictionary

Now, create a nested dictionary like

    {year1: {"winner":value1, "country":value2, "time":value3, "note":value4},
     year2: {"winner":value1, "country":value2, "time":value3, "note":value4},
     ...}
     
*Hint: Start with an empty dictionary. Use a nested for loop.

*Hint 2: Some structure for the nested loop is already given below. Make sure to create a new empty dictionary for each inner dictionary and not just overwrite the keys. More explanations on why this is important will be given in the solutions.

In [ ]:
winner_dict = {}

for item in winners:
    item_dict = {}
    # ...

The first part of the resulting dictionary is

    {'1971': {'winner': 'Beth Bonner',
              'country': 'United States',
              'time': '2:55:22',
              'note': 'World record'},
     '1972': {'winner': 'Nina Kuscsik',
              'country': 'United States',
              'time': '3:08:41',
              'note': ''},
     '1973': {'winner': 'Nina Kuscsik',
              'country': 'United States',
              'time': '2:57:07',
              'note': 'Second victory'},
     '1974': {'winner': 'Kathrine Switzer',
              'country': 'United States',
              'time': '3:07:29',
              'note': ''},
       ... }